ee/backend/app/src/main/kotlin/io/tolgee/ee/service/TaskService.kt

```diff
@@ -0,0 +1,509 @@
+    } else {
+      task.closedAt = Date()
+      task.state = state
+    }
+    taskRepository.saveAndFlush(task)
+    return getTask(projectEntity, taskNumber)
+  }
+
+  @Transactional
+  fun getTask(
+    projectEntity: Project,
+    taskNumber: Long,
+  ): TaskWithScopeView {
+    val task =
+      taskRepository.findByNumber(projectEntity.id, taskNumber).or {
+        throw NotFoundException(Message.TASK_NOT_FOUND)
+      }.get()
+    return getTasksWithScope(listOf(task)).first()
+  }
+
+  @Transactional
+  fun updateTaskKeys(
+    projectEntity: Project,
+    taskNumber: Long,
+    dto: UpdateTaskKeysRequest,
+  ) {
+    val task =
+      taskRepository.findByNumber(projectEntity.id, taskNumber).or {
+        throw NotFoundException(Message.TASK_NOT_FOUND)
+      }.get()
+
+    dto.removeKeys?.let { toRemove ->
+      val taskKeysToRemove =
+        task.keys.filter {
+          toRemove.contains(
+            it.key.id,
+          )
+        }.toMutableSet()
+      task.keys = task.keys.subtract(taskKeysToRemove).toMutableSet()
+      taskKeyRepository.deleteAll(taskKeysToRemove)
+    }
+
+    dto.addKeys?.let { toAdd ->
+      val existingKeys = task.keys.map { it.key.id }.toMutableSet()
+      val nonExistingKeyIds = toAdd.subtract(existingKeys).toMutableSet()
+      val taskKeysToAdd =
+        toAdd
+          .filter { nonExistingKeyIds.contains(it) }
+          .map { TaskKey(task, entityManager.getReference(Key::class.java, it)) }
+      task.keys = task.keys.union(taskKeysToAdd).toMutableSet()
+      taskKeyRepository.saveAll(taskKeysToAdd)
+    }
+  }
+
+  @Transactional
+  fun updateTaskKey(
+    projectEntity: Project,
+    taskNumber: Long,
+    keyId: Long,
+    dto: UpdateTaskKeyRequest,
+  ): UpdateTaskKeyResponse {
+    val task =
+      taskRepository.findByNumber(projectEntity.id, taskNumber).or {
+        throw NotFoundException(Message.TASK_NOT_FOUND)
+      }.get()
+
+    if (task.state == TaskState.CLOSED || task.state == TaskState.DONE) {
+      throw BadRequestException(Message.TASK_NOT_OPEN)
+    }
+
+    val taskWithScope = getTasksWithScope(listOf(task)).first()
+
+    val taskKey =
+      taskKeyRepository.findById(
+        TaskKeyId(
+          task = task,
+          key = entityManager.getReference(Key::class.java, keyId),
+        ),
+      ).or {
+        throw NotFoundException(Message.TASK_NOT_FOUND)
+      }.get()
+
+    val previousValue = taskKey.done
+    val changed = previousValue != dto.done
+
+    val totalDone = taskWithScope.doneItems + if (dto.done) 1 else -1
+    val allDone = totalDone == taskWithScope.totalItems
+    if (changed) {
+      if (dto.done == true) {
+        taskKey.author =
+          entityManager.getReference(
+            UserAccount::class.java,
+            authenticationFacade.authenticatedUser.id,
+          )
+      } else {
+        taskKey.author = null
+      }
+      taskKey.done = dto.done
+      taskKeyRepository.save(taskKey)
+      if (totalDone == 0L) {
```

Comment:
The same would work here `getStateByTotalDone`

Reply:
I'm not really fan of these "universal" rules, you always find cases where it doesn't make sense.

This condition is very simple to read and more understandable than separate method.